File: python/cedegrid/worker.py
"""Credential-free native worker protocol and operator download file helpers."""
from __future__ import annotations

import hashlib
import json
import os
from pathlib import Path
import stat
import uuid


from .errors import CedeGridError

class DrainRequested(CedeGridError):
    code = "ERR_CEDEGRID_DRAIN_REQUESTED"
    """Raised only at a workload-declared safe cancellation boundary."""


def canonical_json(value) -> bytes:
    from .codec import stringify_json
    return stringify_json(value, sort_keys=True).encode()


def sha256_file(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as stream:
        for chunk in iter(lambda: stream.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def sync_directory(path: Path) -> None:
    if os.name == "nt":
        raise NotImplementedError("strict directory durability is unavailable on Windows")
    fd = os.open(path, os.O_RDONLY | getattr(os, "O_DIRECTORY", 0))
    try:
        os.fsync(fd)
    finally:
        os.close(fd)


def durable_mkdir(path: Path, *, durability="strict") -> None:
    missing = []
    cursor = path
    while not cursor.exists():
        missing.append(cursor)
        cursor = cursor.parent
    for directory in reversed(missing):
        directory.mkdir(exist_ok=True)
        if durability == "strict":
            sync_directory(directory)
            sync_directory(directory.parent)


def durable_publish(temporary: Path, final: Path, *, replace: bool = False) -> None:
    if replace:
        os.replace(temporary, final)
    else:
        try:
            os.link(temporary, final)
        except FileExistsError:
            if final.is_symlink() or sha256_file(final) != sha256_file(temporary):
                raise ValueError(f"immutable publication conflict: {final.name}")
        temporary.unlink()
    sync_directory(final.parent)


def atomic_json(path: Path, value: dict, *, replace: bool = False) -> None:
    durable_mkdir(path.parent)
    temporary = path.with_name("." + path.name + "." + uuid.uuid4().hex + ".tmp")
    try:
        with temporary.open("xb") as stream:
            stream.write(canonical_json(value))
            stream.flush()
            os.fsync(stream.fileno())
        durable_publish(temporary, path, replace=replace)
    finally:
        temporary.unlink(missing_ok=True)


class WorkerContext:
    """In-process client of the native supervisor; no authoritative spool writes."""
    def __init__(self, context: dict, drain_file: Path | None = None, *, client=None):
        from .process import _SupervisorClient
        from .codec import integer, I64_MAX
        from .errors import ValidationError
        if context.get("schema_version") != 2:
            raise ValidationError("worker context version 2 is required")
        self.context = dict(context)
        self.context["generation"] = integer(context.get("generation"), "generation", 1, I64_MAX)
        for key in ("namespace_id", "session_id", "task_id", "assignment_id"):
            if type(context.get(key)) is not str or not context[key]:
                raise ValidationError(f"worker context missing {key}")
        self._client = client or _SupervisorClient.from_env(context)
        self.output = Path(context["output_dir"]) if context.get("output_dir") else None
        self.drain_file = drain_file or (Path(context["drain_file"]) if context.get("drain_file") else None)
        self._drain_hooks, self._drain_notified = [], False
        self._artifacts = {}

    @classmethod
    def from_env(cls):
        from .codec import parse_json
        from .errors import ValidationError
        raw = os.environ.get("CEDEGRID_CONTEXT")
        if not raw:
            raise ValidationError("CEDEGRID_CONTEXT is required")
        return cls(parse_json(Path(raw).read_bytes()), Path(os.environ["CEDEGRID_DRAIN_FILE"]) if os.environ.get("CEDEGRID_DRAIN_FILE") else None)

    @property
    def identity(self):
        return {key: self.context[key] for key in ("namespace_id", "session_id", "task_id", "assignment_id", "generation")}

    @property
    def resume(self):
        return self.context.get("resume")

    @property
    def inputs(self):
        return list(self.context.get("inputs", []))

    def on_drain(self, callback):
        if not callable(callback):
            raise ValueError("drain callback must be callable")
        self._drain_hooks.append(callback)

    def draining(self):
        requested = self.drain_file is not None and self.drain_file.exists()
        if requested and not self._drain_notified:
            self._drain_notified = True
            for callback in self._drain_hooks: callback(self)
        return requested

    def safe_point(self):
        if self.draining(): raise DrainRequested("supervisor requested cooperative drain")

    def artifact(self, name, path, *, request_id=None):
        from .errors import ValidationError, RemoteError
        if type(name) is not str or not name or len(name.encode()) > 128 or name in (".", "..") or any(c in name for c in ("/", "\\", "\x00")):
            raise ValidationError("artifact name must be one path component")
        path = Path(path)
        with path.open("rb") as source:
            info = os.fstat(source.fileno())
            if not stat.S_ISREG(info.st_mode): raise ValidationError("artifact source must be a regular file")
            begin = self._client.request("artifact_begin", name=name, size=info.st_size, request_id=request_id or str(uuid.uuid4()))
            upload_id, offset = begin["upload_id"], begin["offset"]
            if not 0 <= offset <= info.st_size: raise RemoteError("invalid artifact offset")
            digest = hashlib.sha256()
            position = 0
            while True:
                chunk = source.read(256 * 1024)
                if not chunk: break
                digest.update(chunk)
                # Replayed begins may acknowledge a durable prefix.
                if position + len(chunk) > offset:
                    tail = chunk[max(0, offset - position):]
                    reply = self._client.request("artifact_chunk", upload_id=upload_id, offset=offset, data_hex=tail.hex())
                    offset += len(tail)
                    if reply.get("offset") != offset: raise RemoteError("artifact offset mismatch")
                position += len(chunk)
            sealed = self._client.request("artifact_finish", upload_id=upload_id)
        if position != info.st_size or sealed.get("size") != info.st_size or sealed.get("sha256") != digest.hexdigest():
            raise RemoteError("artifact integrity failure")
        artifact = {"name": name, **{key: sealed[key] for key in ("artifact_id", "sha256", "size")}}
        self._artifacts[artifact["artifact_id"]] = dict(artifact)
        return artifact

    def _publish(self, kind, metadata, artifacts, publication_id):
        from .codec import stringify_json
        from .errors import PublicationUncertain, RemoteError, ValidationError
        from .process import _InvalidReply
        artifacts = artifacts or []
        if len({item["name"] for item in artifacts}) != len(artifacts): raise ValidationError("artifact names must be unique")
        ids = []
        for item in artifacts:
            if self._artifacts.get(item.get("artifact_id")) != item: raise ValidationError("artifact is not sealed by this worker context")
            ids.append(item["artifact_id"])
        publication_id = publication_id or str(uuid.uuid4())
        fields = {"publication_id": publication_id, "kind": kind, "metadata": metadata, "artifact_ids": ids}
        digest = hashlib.sha256(stringify_json(fields, sort_keys=True).encode()).hexdigest()
        known_digest = None
        try:
            result = self._client.request("publication_commit", request_id=publication_id, **fields)
            known_digest = result.get("sha256")
            if result.get("publication_id") != publication_id or result.get("state") not in ("committed", "rejected"):
                raise _InvalidReply("publication durability acknowledgement missing")
            if result["state"] == "committed" and (result.get("assurance") not in ("durable_local", "replayable_local") or type(known_digest) is not str or len(known_digest) != 64):
                raise _InvalidReply("publication commit evidence missing")
        except RemoteError as error:
            if error.code != "ERR_CEDEGRID_PUBLICATION_UNCERTAIN": raise
            raise PublicationUncertain(publication_id, self.identity, digest=getattr(error, "details", {}).get("sha256"), request_digest=digest, cause=error) from error
        except (OSError, _InvalidReply) as error:
            raise PublicationUncertain(publication_id, self.identity, digest=known_digest, request_digest=digest, cause=error) from error
        if result["state"] == "rejected": raise RemoteError(result.get("message", "publication rejected"))
        return result

    def checkpoint(self, metadata, artifacts=None, *, publication_id=None):
        return self._publish("checkpoint", metadata, artifacts, publication_id)

    def complete(self, metadata, artifacts=None, *, publication_id=None):
        return self._publish("result", metadata, artifacts, publication_id)

    def publication_status(self, publication_id):
        return self._client.request("publication_status", publication_id=publication_id)

    def publication_abort(self, publication_id):
        return self._client.request("publication_abort", publication_id=publication_id)

    def spawn_managed(self, argv, **kwargs):
        from .process import spawn_managed
        self.safe_point()
        return spawn_managed(argv, **kwargs)

File: npm/src/worker.ts
import { createConnection } from 'node:net';
import { readFileSync, existsSync, promises as fs } from 'node:fs';
import { randomUUID, createHash } from 'node:crypto';
import { setTimeout as delay } from 'node:timers/promises';
import { performance } from 'node:perf_hooks';
import { parseJson, stringifyJson, integer, I64_MAX, JsonValue } from './codec';
import { argvField, textField, RpcReply } from './client';
import { CedeGridError, RemoteError, ValidationError, DeadlineExceeded, PublicationUncertain, SpawnUncertain, DrainRequested } from './errors';
export interface AttemptIdentity {
    namespace_id: string;
    session_id: string;
    assignment_id: string;
    generation: bigint;
}
export interface WorkerDescriptor extends AttemptIdentity {
    schema_version: bigint;
    task_id: string;
    output_dir?: string;
    drain_file?: string;
    inputs?: JsonValue[];
    resume?: JsonValue;
}
export interface SealedArtifact {
    name: string;
    artifact_id: string;
    sha256: string;
    size: bigint;
}
export interface PublicationHandle {
    publication_id: string;
    state: 'committed' | 'pending' | 'uncertain' | 'rejected';
    kind: 'checkpoint' | 'result';
    sequence?: bigint;
    sha256?: string;
    assurance?: 'durable_local' | 'replayable_local';
}
export interface SupervisorTransport {
    request(op: string, fields?: Record<string, unknown>): Promise<RpcReply>;
}
class ProtocolError extends CedeGridError {
    constructor(message: string, options?: ErrorOptions) { super(message, 'ERR_CEDEGRID_PROTOCOL', options); }
}
export class SupervisorClient implements SupervisorTransport {
    constructor(readonly endpoint: string, private readonly token: string, readonly identity: AttemptIdentity, readonly timeoutSeconds = 5) { }
    static fromEnv(context?: WorkerDescriptor): SupervisorClient {
        const endpoint = process.env.CEDEGRID_SUPERVISOR_SOCKET, token = process.env.CEDEGRID_SUPERVISOR_TOKEN;
        if (!endpoint || !token)
            throw new ValidationError('assignment does not expose a native supervisor endpoint');
        if (!context) {
            if (process.env.CEDEGRID_CONTEXT) context = readContext();
            else {
                const rawGeneration = process.env.CEDEGRID_ATTEMPT_GENERATION ?? '';
                if (!/^[0-9]+$/.test(rawGeneration)) throw new ValidationError('native supervisor attempt generation is required');
                context = {schema_version: 2n, namespace_id: process.env.CEDEGRID_NAMESPACE_ID!, session_id: process.env.CEDEGRID_SESSION_ID!, assignment_id: process.env.CEDEGRID_ASSIGNMENT_ID!, task_id: process.env.CEDEGRID_TASK_ID!, generation: integer(BigInt(rawGeneration), 'generation', 1n, I64_MAX)};
            }
        }
        for (const key of ['namespace_id', 'session_id', 'assignment_id'] as const) textField(context[key], key, true);
        const { namespace_id, session_id, assignment_id, generation } = context;
        return new SupervisorClient(endpoint, token, { namespace_id, session_id, assignment_id, generation });
    }
    async request(op: string, fields: Record<string, unknown> = {}): Promise<RpcReply> {
        const { request_id = randomUUID(), ...payload } = fields;
        const body = Buffer.from(stringifyJson({ version: 2n, op, request_id, token: this.token, ...this.identity, ...payload }) + '\n');
        const limit = 3 * 1024 * 1024;
        if (body.length > limit)
            throw new ValidationError('supervisor request exceeds 3 MiB');
        return new Promise((resolve, reject) => {
            const socket = createConnection(this.endpoint);
            let chunks: Buffer[] = [];
            let size = 0;
            let finished = false;
            const finish = (error?: unknown, value?: RpcReply) => { if (finished)
                return; finished = true; clearTimeout(timer); socket.destroy(); error ? reject(error) : resolve(value!); };
            const timer = setTimeout(() => finish(new DeadlineExceeded('supervisor request deadline exceeded')), this.timeoutSeconds * 1000);
            socket.on('connect', () => socket.end(body));
            socket.on('error', error => finish(error));
            socket.on('data', (chunk: Buffer) => {
                chunks.push(chunk);
                size += chunk.length;
                if (size > limit) {
                    finish(new ProtocolError('supervisor reply exceeds 3 MiB'));
                    return;
                }
                if (chunk.includes(10)) {
                    try {
                        const bytes = Buffer.concat(chunks);
                        if (bytes[bytes.length - 1] !== 10)
                            throw new ProtocolError('invalid supervisor frame');
                        const reply = parseJson(new TextDecoder('utf-8', { fatal: true }).decode(bytes));
                        if (!reply || typeof reply !== 'object' || Array.isArray(reply) || typeof reply.ok !== 'boolean')
                            throw new ProtocolError('invalid supervisor reply');
                        if (!reply.ok)
                            throw Object.assign(new RemoteError(typeof reply.message === 'string' ? reply.message : 'supervisor rejected request', typeof reply.code === 'string' ? reply.code : undefined), { details: reply });
                        finish(undefined, reply);
                    }
                    catch (cause) {
                        finish(cause instanceof RemoteError ? cause : new ProtocolError('invalid supervisor reply', { cause }));
                    }
                }
            });
            socket.on('end', () => { if (!finished)
                finish(new ProtocolError('supervisor acknowledgement missing')); });
        });
    }
}
function readContext(): WorkerDescriptor {
    const path = process.env.CEDEGRID_CONTEXT;
    if (!path)
        throw new ValidationError('CEDEGRID_CONTEXT is required');
    const context = parseJson(readFileSync(path, 'utf8')) as unknown as WorkerDescriptor;
    if (!context || context.schema_version !== 2n)
        throw new ValidationError('worker context version 2 is required');
    for (const key of ['namespace_id', 'session_id', 'assignment_id', 'task_id'] as const)
        textField(context[key], key, true);
    context.generation = integer(context.generation, 'generation', 1n, I64_MAX);
    return context;
}
export interface SpawnOptions {
    cwd?: string;
    env?: Record<string, string>;
    noEscape: boolean;
    singleProcess: boolean;
    requestId?: string;
}
export class ManagedChild {
    constructor(private readonly client: SupervisorTransport, readonly childId: string) { }
    status(): Promise<RpcReply> { return this.client.request('status', { child_id: this.childId }); }
    stop(): Promise<RpcReply> { return this.client.request('stop', { child_id: this.childId }); }
    async wait(options: {
        timeoutSeconds?: number;
        pollIntervalSeconds?: number;
    } = {}): Promise<RpcReply> {
        const interval = options.pollIntervalSeconds ?? 0.1;
        if (!Number.isFinite(interval) || interval <= 0 || options.timeoutSeconds !== undefined && (!Number.isFinite(options.timeoutSeconds) || options.timeoutSeconds < 0))
            throw new ValidationError('invalid child wait intervals');
        const deadline = options.timeoutSeconds === undefined ? Infinity : performance.now() + options.timeoutSeconds * 1000;
        for (;;) {
            const reply = await this.status();
            if (reply.state === 'released')
                return reply;
            if (reply.state === 'needs_reconciliation')
                throw new RemoteError('child remains reserved and requires reconciliation', 'ERR_CEDEGRID_RECONCILIATION_REQUIRED');
            if (performance.now() >= deadline)
                throw new DeadlineExceeded('child wait expired; child was not signaled');
            await delay(Math.min(interval * 1000, deadline - performance.now()));
        }
    }
}
export async function spawnManaged(argv: string[], options: SpawnOptions): Promise<ManagedChild> {
    argvField(argv);
    if (options.noEscape !== true || options.singleProcess !== true)
        throw new ValidationError('explicit singleProcess/noEscape acknowledgement required');
    if (options.cwd !== undefined && !/^(\/|[A-Za-z]:[\\/]|\\\\)/.test(options.cwd))
        throw new ValidationError('child cwd must be absolute');
    const client = SupervisorClient.fromEnv(), requestId = options.requestId ?? randomUUID();
    const fields: Record<string, unknown> = { request_id: requestId, argv, no_escape: true, single_process: true };
    if (options.cwd !== undefined)
        fields.cwd = options.cwd;
    if (options.env !== undefined)
        fields.env = options.env;
    try {
        const reply = await client.request('spawn', fields);
        if (typeof reply.child_id !== 'string' || !reply.child_id)
            throw new ProtocolError('spawn reply missing child_id');
        return new ManagedChild(client, reply.child_id);
    }
    catch (cause) {
        if (cause instanceof RemoteError)
            throw cause;
        throw new SpawnUncertain(requestId, { cause });
    }
}
export class WorkerContext {
    private readonly client: SupervisorTransport;
    private readonly artifacts = new Map<string, SealedArtifact>();
    private readonly drainHooks: Array<(context: WorkerContext) => void> = [];
    private drainNotified = false;
    constructor(readonly context: WorkerDescriptor, client?: SupervisorTransport) {
        if (context.schema_version !== 2n)
            throw new ValidationError('worker context version 2 is required');
        for (const key of ['namespace_id', 'session_id', 'assignment_id', 'task_id'] as const)
            textField(context[key], key, true);
        integer(context.generation, 'generation', 1n, I64_MAX);
        this.client = client ?? SupervisorClient.fromEnv(context);
    }
    static fromEnv(): WorkerContext { return new WorkerContext(readContext()); }
    get identity(): AttemptIdentity & {
        task_id: string;
    } { const { namespace_id, session_id, assignment_id, generation, task_id } = this.context; return { namespace_id, session_id, assignment_id, generation, task_id }; }
    get inputs(): JsonValue[] { return [...(this.context.inputs ?? [])]; }
    get resume(): JsonValue | undefined { return this.context.resume; }
    onDrain(callback: (context: WorkerContext) => void): void { this.drainHooks.push(callback); }
    draining(): boolean { const path = this.context.drain_file ?? process.env.CEDEGRID_DRAIN_FILE; const requested = !!path && existsSync(path); if (requested && !this.drainNotified) {
        this.drainNotified = true;
        for (const callback of this.drainHooks)
            callback(this);
    } return requested; }
    safePoint(): void { if (this.draining())
        throw new DrainRequested(); }
    async artifact(name: string, path: string, options: {
        requestId?: string;
    } = {}): Promise<SealedArtifact> {
        textField(name, 'artifact name', true);
        if (Buffer.byteLength(name) > 128 || name === '.' || name === '..' || /[\\/]/.test(name))
            throw new ValidationError('artifact name must be one path component');
        const source = await fs.open(path, 'r');
        try {
            const info = await source.stat({ bigint: true });
            if (!info.isFile())
                throw new ValidationError('artifact source must be a regular file');
            const begin = await this.client.request('artifact_begin', { request_id: options.requestId ?? randomUUID(), name, size: info.size });
            let offset = integer(begin.offset, 'artifact offset', 0n, info.size);
            const uploadId = begin.upload_id;
            const hash = createHash('sha256');
            const buffer = Buffer.alloc(256 * 1024);
            let position = 0n;
            for (;;) {
                const { bytesRead } = await source.read(buffer, 0, buffer.length, null);
                if (!bytesRead)
                    break;
                const chunk = buffer.subarray(0, bytesRead);
                hash.update(chunk);
                if (position + BigInt(bytesRead) > offset) {
                    const tail = chunk.subarray(Number(offset > position ? offset - position : 0n));
                    const reply = await this.client.request('artifact_chunk', { upload_id: uploadId, offset, data_hex: tail.toString('hex') });
                    offset += BigInt(tail.length);
                    if (reply.offset !== offset)
                        throw new ProtocolError('artifact offset mismatch');
                }
                position += BigInt(bytesRead);
            }
            const sealed = await this.client.request('artifact_finish', { upload_id: uploadId });
            if (position !== info.size || sealed.size !== info.size || sealed.sha256 !== hash.digest('hex') || typeof sealed.artifact_id !== 'string')
                throw new ProtocolError('artifact integrity failure');
            const artifact: SealedArtifact = { name, artifact_id: sealed.artifact_id, sha256: sealed.sha256, size: sealed.size };
            this.artifacts.set(artifact.artifact_id, { ...artifact });
            return artifact;
        }
        finally {
            await source.close();
        }
    }
    private async publish(kind: 'checkpoint' | 'result', metadata: JsonValue, artifacts: SealedArtifact[], publicationId: string = randomUUID()): Promise<PublicationHandle> {
        if (new Set(artifacts.map(a => a.name)).size !== artifacts.length)
            throw new ValidationError('artifact names must be unique');
        for (const artifact of artifacts)
            if (stringifyJson(this.artifacts.get(artifact.artifact_id) ?? null) !== stringifyJson(artifact))
                throw new ValidationError('artifact is not sealed by this context');
        const fields = { publication_id: publicationId, kind, metadata, artifact_ids: artifacts.map(a => a.artifact_id) };
        const digest = createHash('sha256').update(stringifyJson(fields)).digest('hex');
        let knownDigest: string | undefined;
        try {
            const reply = await this.client.request('publication_commit', { request_id: publicationId, ...fields });
            knownDigest = typeof reply.sha256 === 'string' ? reply.sha256 : undefined;
            if (reply.publication_id !== publicationId || !['committed', 'rejected'].includes(reply.state))
                throw new ProtocolError('publication durability acknowledgement missing');
            if (reply.state === 'committed' && (!['durable_local', 'replayable_local'].includes(reply.assurance) || !knownDigest || !/^[a-f0-9]{64}$/.test(knownDigest)))
                throw new ProtocolError('publication commit evidence missing');
            if (reply.state === 'rejected')
                throw new RemoteError(reply.message ?? 'publication rejected');
            return reply as PublicationHandle;
        }
        catch (cause) {
            if (cause instanceof RemoteError && cause.code !== 'ERR_CEDEGRID_PUBLICATION_UNCERTAIN')
                throw cause;
            if (cause instanceof RemoteError && 'details' in cause)
                knownDigest = (cause.details as any)?.sha256;
            throw new PublicationUncertain(publicationId, { ...this.identity }, knownDigest, { cause }, digest);
        }
    }
    checkpoint(metadata: JsonValue, artifacts: SealedArtifact[] = [], options: {
        publicationId?: string;
    } = {}): Promise<PublicationHandle> { return this.publish('checkpoint', metadata, artifacts, options.publicationId); }
    complete(metadata: JsonValue, artifacts: SealedArtifact[] = [], options: {
        publicationId?: string;
    } = {}): Promise<PublicationHandle> { return this.publish('result', metadata, artifacts, options.publicationId); }
    async publicationStatus(publicationId: string): Promise<PublicationHandle> { return await this.client.request('publication_status', { publication_id: publicationId }) as PublicationHandle; }
    async publicationAbort(publicationId: string): Promise<PublicationHandle> { return await this.client.request('publication_abort', { publication_id: publicationId }) as PublicationHandle; }
    spawnManaged(argv: string[], options: SpawnOptions): Promise<ManagedChild> { this.safePoint(); return spawnManaged(argv, options); }
}
